src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -1090,23 +1095,99 @@ bool G1CollectedHeap::expand_single_region(uint node_index) {
          shrink_bytes, G1HeapRegion::GrainBytes);
 
   uint num_regions_to_remove = (uint)(shrink_bytes / G1HeapRegion::GrainBytes);
+  uint num_regions_removed = 0;
+
+  // Use time-based selection to shrink oldest eligible regions
+  log_debug(gc, ergo, heap)("Time-based shrink: removing %u oldest regions (%zuB)",
+                            num_regions_to_remove, shrink_bytes);
+  num_regions_removed = _hrm.shrink_by(num_regions_to_remove, true /* use_time_based_selection */);
 
-  uint num_regions_removed = _hrm.shrink_by(num_regions_to_remove);
   size_t shrunk_bytes = num_regions_removed * G1HeapRegion::GrainBytes;
+  log_debug(gc, ergo, heap)("Time-based shrink: Requested shrinking amount: %zuB actual shrinking amount: %zuB (%u regions)",
+                           shrink_bytes, shrunk_bytes, num_regions_removed);
+
+  if (num_regions_removed > 0) {
+    log_info(gc, heap)("Time-based shrink: uncommitted %u oldest regions (%zuMB), heap size now %zuMB",
+                       num_regions_removed, shrunk_bytes / M, capacity() / M);
+    log_debug(gc, heap)("Time-based shrink details: requested=%zuB actual=%zuB "
+                        "regions_removed=%u heap_capacity=%zuB",
+                        shrink_bytes, shrunk_bytes, num_regions_removed, capacity());
+    policy()->record_new_heap_size(num_committed_regions());
+  } else {
+    log_debug(gc, ergo, heap)("Time-based shrink: Did not shrink the heap (no eligible regions found)");
+  }
+}
+
+void G1CollectedHeap::shrink_helper(size_t shrink_bytes) {
+  assert(shrink_bytes > 0, "must be");
+  assert(is_aligned(shrink_bytes, G1HeapRegion::GrainBytes),
+         "Shrink request for %zuB not aligned to heap region size %zuB",
+         shrink_bytes, G1HeapRegion::GrainBytes);
+
+  uint num_regions_to_remove = (uint)(shrink_bytes / G1HeapRegion::GrainBytes);
+  uint num_regions_removed = 0;
 
+  // Always perform normal heap shrinking when requested
+  // This preserves the original GC-triggered shrinking behavior
+  log_debug(gc, ergo, heap)("Heap shrink requested: removing %u regions (%zuB)",
+                            num_regions_to_remove, shrink_bytes);
+  num_regions_removed = _hrm.shrink_by(num_regions_to_remove);
+
+  size_t shrunk_bytes = num_regions_removed * G1HeapRegion::GrainBytes;
   log_debug(gc, ergo, heap)("Heap resize. Requested shrinking amount: %zuB actual shrinking amount: %zuB (%u regions)",
                             shrink_bytes, shrunk_bytes, num_regions_removed);
+
   if (num_regions_removed > 0) {
+    log_info(gc, heap)("Heap shrink details: uncommitted %u regions (%zuMB), heap size now %zuMB",
+                       num_regions_removed, shrunk_bytes / M, capacity() / M);
+    log_debug(gc, heap)("Heap shrink details: requested=%zuB actual=%zuB "
+                        "regions_removed=%u heap_capacity=%zuB",
+                        shrink_bytes, shrunk_bytes, num_regions_removed, capacity());
```

Comment:
It would be nice to not repeat the same information at different levels; one can check the current level with `LogTarget::is_enabled()`.